File: leetcode/112/PathSum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {boolean}
 */
var hasPathSum = function(root, sum) {
    if(!root) {
        return false;
    }
    if(!root.left && !root.right) {
        return root.val == sum ? true : false;
    }
    let remain = sum - root.val
    return hasPathSum(root.left, remain) || hasPathSum(root.right, remain)
};

File: leetcode/112/PathSum.java
import java.util.*;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if(root==null)
            return false;

        if(root.left == null && root.right == null && sum - root.val == 0)
            return true;
        else
            return hasPathSum(root.left, sum - root.val) || hasPathSum(root.right, sum - root.val);
    }
}

class Solution2 {
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null)
            return false;
        Stack<TreeNode> stackNode = new Stack<TreeNode>();
        Stack<Integer> stackSum = new Stack<Integer>();

        stackNode.add(root);
        stackSum.add(sum - root.val);

        while (!stackNode.empty()) {
            TreeNode cNode = stackNode.pop();
            int cSum = stackSum.pop();

            if (cNode.left == null && cNode.right == null && cSum == 0)
                return true;

            if (cNode.left != null) {
                stackNode.add(cNode.left);
                stackSum.add(cSum - cNode.left.val);
            }

            if (cNode.right != null) {
                stackNode.add(cNode.right);
                stackSum.add(cSum - cNode.right.val);
            }
        }

        return false;
    }
}